dataRetriever/resolvers/equivalentProofsResolver.go

```diff
@@ -0,0 +1,293 @@
+	EquivalentProofsNonceHashStorage storage.Storer
+	EquivalentProofsPool             processor.EquivalentProofsPool
+	IsFullHistoryNode                bool
+}
+
+type equivalentProofsResolver struct {
+	*baseResolver
+	baseStorageResolver
+	messageProcessor
+	dataPacker                       dataRetriever.DataPacker
+	equivalentProofsNonceHashStorage storage.Storer
+	equivalentProofsPool             processor.EquivalentProofsPool
+}
+
+// NewEquivalentProofsResolver creates an equivalent proofs resolver
+func NewEquivalentProofsResolver(args ArgEquivalentProofsResolver) (*equivalentProofsResolver, error) {
+	err := checkArgEquivalentProofsResolver(args)
+	if err != nil {
+		return nil, err
+	}
+
+	return &equivalentProofsResolver{
+		baseResolver: &baseResolver{
+			TopicResolverSender: args.SenderResolver,
+		},
+		baseStorageResolver: createBaseStorageResolver(args.EquivalentProofsStorage, args.IsFullHistoryNode),
+		messageProcessor: messageProcessor{
+			marshalizer:      args.Marshaller,
+			antifloodHandler: args.AntifloodHandler,
+			throttler:        args.Throttler,
+			topic:            args.SenderResolver.RequestTopic(),
+		},
+		dataPacker:                       args.DataPacker,
+		equivalentProofsNonceHashStorage: args.EquivalentProofsNonceHashStorage,
+		equivalentProofsPool:             args.EquivalentProofsPool,
+	}, nil
+}
+
+func checkArgEquivalentProofsResolver(args ArgEquivalentProofsResolver) error {
+	err := checkArgBase(args.ArgBaseResolver)
+	if err != nil {
+		return err
+	}
+	if check.IfNil(args.DataPacker) {
+		return dataRetriever.ErrNilDataPacker
+	}
+	if check.IfNil(args.EquivalentProofsStorage) {
+		return fmt.Errorf("%w for EquivalentProofsStorage", dataRetriever.ErrNilProofsStorage)
+	}
+	if check.IfNil(args.EquivalentProofsNonceHashStorage) {
+		return fmt.Errorf("%w for EquivalentProofsNonceHashStorage", dataRetriever.ErrNilProofsStorage)
+	}
+	if check.IfNil(args.EquivalentProofsPool) {
+		return dataRetriever.ErrNilProofsPool
+	}
+
+	return nil
+}
+
+// ProcessReceivedMessage represents the callback func from the p2p.Messenger that is called each time a new message is received
+// (for the topic this validator was registered to, usually a request topic)
+func (res *equivalentProofsResolver) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedPeer core.PeerID, source p2p.MessageHandler) ([]byte, error) {
+	err := res.canProcessMessage(message, fromConnectedPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	res.throttler.StartProcessing()
+	defer res.throttler.EndProcessing()
+
+	rd, err := res.parseReceivedMessage(message, fromConnectedPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	switch rd.Type {
+	case dataRetriever.HashType:
+		return nil, res.resolveHashRequest(rd.Value, rd.Epoch, message.Peer(), source)
+	case dataRetriever.HashArrayType:
+		return nil, res.resolveMultipleHashesRequest(rd.Value, rd.Epoch, message.Peer(), source)
+	case dataRetriever.NonceType:
```

Comment:
do we also have a nonce array type?

Reply:
no, might not be needed for proofs either